### PR TITLE
Generate unions of shapes

### PR DESCRIPTION
**Pull request by ianhoffman**
This PR allows Hack-JSON-Schema to generate "best effort" unions from Hack shapes.

Hack doesn't actually support unions, so there's no way to write `T1 = T2 | T3 | T4`. On the other hand, we can do a best effort impression of a union by deriving the "lowest upper bound" type for T2, T3, and T4. [Prior work](https://github.com/slackhq/hack-json-schema/pull/67) has introduced support for "best effort" unions for primitive types, vectors and keysets.

Generating "best effort" unions for Hack shapes is the largest piece of unfinished work in this effort. The following rules apply:

```hack
// If any shape in the union is open, the generated shape must be open
t1 = shape();
t2 = shape(...);
t3 = shape(...); // union of t1 & t2

// If any field is absent from any other shape, then the field must be optional in the generated shape
t1 = shape('foo' => int);
t2 = shape('bar' => str);
t3 = shape(?'foo' => int, ?'bar' => str); // union of t1 & t2

// If a field is present in all shapes, then its type is the union of the type of the field across all shapes
t1 = shape('foo' => int);
t2 = shape('foo' => str);
t3 = shape('foo' => arraykey); // union of t1 & t2

// If a field is optional in one shape, it will be optional in the generated shape
t1 = shape(?'foo' => int);
t2 = shape('foo' => int);
t3 = shape(?'foo' => int); // union of t1 and t2

// If a field is absent in an open shape, the field will be absent from the generated shape
// since it would be typed as `?'foo' => mixed` which is pointless
t1 = shape('foo' => int);
t2 = shape(...);
t3 = shape(...); // union of t1 and t2

// If a field unions to `mixed` and is optional, it will be absent from the generated shape
// since it would be typed as `?'foo' => mixed` which is pointless
t1 = shape(?'foo' => ?int, ...);
t2 = shape(?'foo' => vec<string>, ...);
t3 = shape(...); // union of t1 and t2
```

Tests are in https://github.com/slackhq/hack-json-schema/pull/70

### CODE REVIEW

File: src/Codegen/TypeRenderer.hack [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+      return $codegen_value;
+    }
+  }
+
+  private function visitOptionalType(Typing\OptionalType $type, string $type_name): string {
+    // Visit the inner type and then generate a type like '?<inner type>'
+    $codegen_type = $this->visitType($type->getType(), $this->generateTypeNameFromParts($type_name, 'nonnull'));
+    switch ($codegen_type) {
+      case Typing\ConcreteTypeName::NONNULL:
+        return 'mixed';
```
Comment (mwildehahn) [inline src/Codegen/TypeRenderer.hack:83]: do you want nonnull here or is mixed intentional?

File: src/Codegen/Typing/TypeSystem.hack [hunk 2]
```diff
@@ -182,10 +180,90 @@ final abstract class TypeSystem {
...
+      $required_field_counts = dict[];
+
+      foreach ($types as $ty) {
+        if ($ty->getConcreteTypeName() === ConcreteTypeName::SHAPE) {
+          $num_shapes++;
+          if ($ty->isClosedShape()) {
+            $num_closed_shapes++;
+          }
+
+          foreach ($ty->getShapeFields() as $name => $field) {
```
Comment (mwildehahn) [inline src/Codegen/Typing/TypeSystem.hack:225]: how does this handle shapes within shapes? these just get listed out as "shape" types?

**Review by mwildehahn — APPROVED** [review]
A couple small questions, but looks good!

File: src/Codegen/TypeRenderer.hack (hunk 1 above)
Comment (ianhoffman) [reply src/Codegen/TypeRenderer.hack:83]: `mixed` is intentional. In this method we're generating code for an optional type, so `?nonnull` should be represented as `mixed`.

File: src/Codegen/Typing/TypeSystem.hack (hunk 2 above)
Comment (ianhoffman) [reply src/Codegen/Typing/TypeSystem.hack:225]: Yup, exactly. If multiple shapes contain a field with the same name, we'll union the types of that field recursively, here: https://github.com/slackhq/hack-json-schema/pull/78/files#diff-18da79e1d9a43454019fd566249550acdfcc9397cd8f5b915ca806ff7f053a44R248

That's how we handle nested shapes as well.

**Review by mwildehahn — APPROVED** [review]

**Pull request merged**